lib/tessel/deploy.js

```diff
@@ -421,20 +440,85 @@ actions.resolveBinaryModules = function(opts) {
       // Gather information about each found module
       var binName = path.basename(mPath);
       var modulePath = bindings.getRoot(mPath);
       var packageJson = require(path.join(globRoot, modulePath, 'package.json'));
 
-      return {
+      if (mPath.endsWith('.gyp')) {
+
+        // Check that there are no matching .node paths in the ALREADY PROCESSED paths
+        for (var i = 0; i < bins.length; i++) {
+          if (bins[i].path === modulePath) {
+            // When the currently resolving binary module path has been found
+            // in the list of existing binary module paths that have already been
+            // accounted for, return immediately to continue to processing the
+            // found binary module path candidates.
+            //
+            // An example:
+            //
+            // [
+            //  "node_modules/bufferutil/build/Release/bufferutil.node",
+            //  "node_modules/bufferutil/binding.gyp",
+            // ]
+            //
+            // Note, the order will always be .node then .gyp
+            return bins;
+          }
+        }
+
+        // If unfound, then we have to dig around in the binding.gyp for the target_name
+        // to figure out the actual name of the .node file
+        //
+        // If you're interested in seeing just how much of a nightmare mess these
+        // files can be, take a look at this:
+        //
+        // https://github.com/nodejs/node-gyp/wiki/%22binding.gyp%22-files-out-in-the-wild
+        //
+        var bindingGypPath = path.join(globRoot, modulePath, binName);
+        var bindingGypData;
+        var bindingGypJson;
+
+        try {
+          // Sometimes people write really tidy binding.gyp files that are
+          // actually valid JSON, which is totally awesome!
+          bindingGypJson = require(bindingGypPath);
+        } catch (error) {
+          // ... Other times binding.gyp is an inconsistent mess, but it's still
+          // a valid Python data structure. So we can spawn a python to read it.
+          // Sounds gross, but there is no other clear way to do this.
+          var bindingGypData = actions.resolveBinaryModules.readGypFileSync(bindingGypPath);
+
+          try {
+            bindingGypJson = JSON.parse(bindingGypData);
+          } catch (error) {
+            // If this module's binding.gyp is missing, broken or otherwise
+            // unusable, log a message about and move on. There are too
+            // many failure modes here, no way to recover.
+
+            logMissingBinaryModuleWarning(packageJson.name);
+            return bins;
+          }
+        }
+
+        if (bindingGypJson && Array.isArray(bindingGypJson.targets)) {
+          // Anything that can't be covered by this will have to be
+          // dealt with as we encounter them and as they are reported.
+          binName = bindingGypJson.targets[0].target_name + '.node';
```

Comment:
Seems reasonable. I don't really understand why there can be multiple targets but that's not really our problem.